xdem/coreg.py

```diff
@@ -67,6 +66,173 @@ def _calculate_slope_and_aspect_nuthkaab(dem: NDArrayf) -> tuple[NDArrayf, NDArr
+
+    return (df_shifted[z_name].values - dem_h) * weight_
+
+def df_sampling_from_dem(dem: NDArrayf, tba_dem: NDArrayf, samples:int = 5000, order:int = 1, offset = None) -> pd.DataFrame:
+    '''
+    generate a datafram from a dem by random sampling.
+
+    :param offset: The pixel’s center is returned by default, but a corner can be returned by setting offset to one of ul, ur, ll, lr.
+    
+    :returns dataframe: N,E coordinates and z of DEM at sampling points.
+    '''
+    try:
+        if offset is None and (dem.tags['AREA_OR_POINT'] in ['Point','point']):
+            offset = 'center'
+        else:
+            offset = 'ul'
+    except KeyError:
+        offset = 'ul'
+
+    # Avoid edge, and mask-out-area in sampling
+    width,length = dem.shape
+    i,j = np.random.randint(10, width-10, samples),np.random.randint(10, length-10, samples)
+    mask = dem.data.mask[0]
+
+    # Get value
+    x, y = dem.ij2xy(i[~mask[i,j]], j[~mask[i,j]], offset = offset)
+    z = scipy.ndimage.map_coordinates(dem.data[0, :, :], [i[~mask[i,j]], j[~mask[i,j]]], order=order, mode="nearest")
+    df = pd.DataFrame({'z': z,'N':y,'E':x})
+
+    # mask out from tba_dem
+    if tba_dem is not None:
+        df, _ = mask_dataframe_by_dem(df,tba_dem)
+
+    return df
+
+def mask_dataframe_by_dem(df: pd.DataFrame or np.ndarray, dem: NDArrayf) -> pd.DataFrame or np.ndarray:
+    '''
+    mask out the dataframe (has 'E','N' columns), or np.ndarray ([E,N])  by DEM's mask
+    return new dataframe and mask
+    '''
+
+    final_mask = ~dem.data.mask
+    mask_raster = dem.copy(new_array = final_mask.astype(np.float32))
+
+    if isinstance(df, pd.DataFrame):
+        pts = np.array((df['E'].values, df['N'].values)).T
+    elif isinstance(df, np.ndarray):
+        pts = df
+
+    ref_inlier = mask_raster.interp_points(pts, input_latlon=False, order=0)
+    new_df = df[ref_inlier.astype(bool)].copy()
+
+    return new_df, ref_inlier.astype(bool)
+
+def fill_by_nearest(data: np.ndarray, invalid = None) -> np.ndarray:
+    """
+    https://stackoverflow.com/questions/3662361/fill-in-missing-values-with-nearest-neighbour-in-python-numpy-masked-arrays
+
+    Replace the value of invalid 'data' cells (indicated by 'invalid') 
+    by the value of the nearest valid data cell
+
+    :param data:    numpy array of any dimension
+    :param invalid: a binary array of same shape as 'data', 'mask'. 
+    True cells set where data value should be replaced. If None (default), use: invalid  = np.isnan(data)
+
+    Return a filled array. 
+    """
+
+    if invalid is None: invalid = np.isnan(data)
+
+    ind = scipy.ndimage.distance_transform_edt(invalid, return_distances=False, return_indices=True)
+    return data[tuple(ind)]
+
+def interp_points(
+    dem: xdem.DEM,
+    pts: np.ndarray,
+    input_latlon: bool = False,
+    mode: str = "nearest",
+    area_or_point: str = 'Area' or None,
+    order=1,
+    fillhole=False,
+    **kwargs: Any,
+    ) -> np.ndarray:
```

Comment:
Please add a docstring!

Also `fillhole` may be better named as `fill_holes`, because I guess we're talking about more than one hole, right?